### PR TITLE
change(web): conversion to standard filesystem layout:  tools -> src/tools 📜

### PR DESCRIPTION
**Pull request by jahorton**
This 📜 PR set is for some long-needed maintenance:  the Web project's files are being reorganized into a format more consistent with the other Keyman projects.  There's no new functionality; the end-goal is for everything to be stable and working as before (at the end of the 📜 PR chain).

Continues from #7515.

This PR moves all "tool" projects designed to assist KMW builds and testing into the src/tools/ folder, each going to an appropriate subdirectory based on its role:
- Tools used during builds:  `src/tools/building`
   - Since we auto-.gitignore `build` folders, something a bit distinct was needed.
- Tools used during tests: `src/tools/testing` 

The `bulk_renderer` tool has also been relocated as a "testing" tool under `src/tools/testing`.

Finally, since some of the former "tools" were merely subsets of the engine code... those tools have been formally reworked _as_ engine submodules, with the main common Web engine under `engine/main`.

The web/ directory's "index.html" infrastructure has also been updated to maintain links among the relevant moving parts, including adding a top-level entry linking to pages for KMW tools when they exist.

Finally, adding a few more-specific builder-based scripts has allowed narrowing the focus on some of the build-script dependency setup.

@keymanapp-test-bot skip

Unit tests have been delegated to #7831.

### CODE REVIEW

File: web/src/engine/build.sh [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+  local COMPILE_TARGET=$1
+  local COMPILED_OUTPUT_PATH=$(output_path $COMPILE_TARGET $OUTPUT_DIR)
+
+  BUNDLE_CONFIG=src/$COMPILE_TARGET/tsconfig.bundled.json
+
+  $compilecmd -b src/$COMPILE_TARGET -v
+
+  # Handle any 'bundled' compilations, too!
+  if [ -f "$BUNDLE_CONFIG" ]; then
```
Comment (jahorton) [inline web/src/engine/build.sh:144]: A deficiency has been noted here address by https://github.com/keymanapp/keyman/pull/7831/commits/9c569b76230be8f3b68c92b617a68f76b37ad904 in the next PR, at least at present.

tl;dr:  I made a bad assumption about the working directory for the if-check due to running in a subdirectory.  The following `$compilecmd` isn't affected b/c it involves an `npm` script call.

File: web/src/engine/build.sh [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+# Definition of global compile constants
+
+MAIN="engine/main" # Covers all engine code, including submodules like those listed below.
+DEVICEDETECT="engine/device-detect"
+ELEMENTWRAPPERS="engine/element-wrappers"
+
+BUILD_BASE="build"
+
+OUTPUT_DIR="obj"
```
Comment (mcdurdin) [inline web/src/engine/build.sh:32]: Given these strings have no spaces, technically none of them need quotes :grin:
```suggestion
MAIN=engine/main     # Covers all engine code, including submodules like those listed below.
DEVICEDETECT=engine/device-detect
ELEMENTWRAPPERS=engine/element-wrappers

BUILD_BASE=build

OUTPUT_DIR=obj
```

File: web/src/engine/build.sh [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+#
+# ### Parameters
+#
+# * 1: - build product (app/embed, app/web, app/ui, engine)
+# * 2: (optional) - build stage / config (obj, debug, release)
+#
+# ### Example
+#
+# ```bash
+#   cp index.js $(output_path app/web debug)/index.js
```
Comment (mcdurdin) [inline web/src/engine/build.sh:45]:
```suggestion
#   cp index.js "$(output_path app/web debug)/index.js"
```

File: web/src/engine/build.sh [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+# ```
+#
+# The block above would copy index.js into the build output folder for app/web's debug
+# product.
+#
+# ``` bash
+#   rm -rf $(output_path app/web)
```
Comment (mcdurdin) [inline web/src/engine/build.sh:52]:
```suggestion
#   rm -rf "$(output_path app/web)"
```

File: web/src/engine/build.sh [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+# (Facilitates automated setup for build agents.)
+PATH="../../../node_modules/.bin:$PATH"
+
+compiler="npm run tsc --"
+compilecmd="$compiler"
+
+builder_check_color "$@"
+
+# TODO: for predictive-text, we only need :headless, perhaps we should be splitting modules?
+# TODO: remove :tools once kmlmc is a dependency for test:module
```
Comment (mcdurdin) [inline web/src/engine/build.sh:87]: Are these TODOs obsolete?

File: web/src/engine/build.sh [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+DOC_WEB_PRODUCT="${BUILDER_TERM_START}:web${BUILDER_TERM_END} build product"
+DOC_TEST_WEB="${BUILDER_TERM_START}test:web${BUILDER_TERM_END}"
+DOC_BUILD_EMBED_WEB="${BUILDER_TERM_START}build:embed${BUILDER_TERM_END} and ${BUILDER_TERM_START}build:web${BUILDER_TERM_END}"
+DOC_TEST_SYMBOL="actions - ${BUILDER_TERM_START}test${BUILDER_TERM_END}"
```
Comment (mcdurdin) [inline web/src/engine/build.sh:92]: Are these needed? If not, then probably `builder_check_color` isn't either

File: web/src/engine/build.sh [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+# ```bash
+#   compile engine/main
+# ```
+compile ( ) {
+  if [ $# -lt 1 ]; then
+    fail "Scripting error: insufficient argument count!"
+  fi
+
+  local COMPILE_TARGET=$1
+  local COMPILED_OUTPUT_PATH=$(output_path $COMPILE_TARGET $OUTPUT_DIR)
```
Comment (mcdurdin) [inline web/src/engine/build.sh:137]:
```suggestion
  local COMPILED_OUTPUT_PATH="$(output_path $COMPILE_TARGET $OUTPUT_DIR)"
```

File: web/src/engine/main/tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  ],
+
+	"exclude": [
+		"./kmwnative.ts",
+		"./kmwembedded.ts",
+		"./kmwuibutton.ts",
+		"./kmwuifloat.ts",
+		"./kmwuitoggle.ts",
+		"./kmwuitoolbar.ts",
+	],
```
Comment (mcdurdin) [inline web/src/engine/main/tsconfig.json:27]: These seem to have some overlap?

File: web/src/tools/building/sourcemap-root/build.sh [hunk 9]
```diff
@@ -36,12 +36,12 @@ if builder_start_action configure; then
...
 
 if builder_start_action clean; then
-  rm -f index.js
+  rm -rf ../../../../build/tools/building/sourcemap-root
   builder_finish_action success clean
 fi
 
 if builder_start_action build; then
-  npm run tsc -- -b tools/sourcemap-root/tsconfig.json
+  npm run tsc -- -b $THIS_SCRIPT_PATH/tsconfig.json
```
Comment (mcdurdin) [inline web/src/tools/building/sourcemap-root/build.sh:44]:
```suggestion
  npm run tsc -- -b "$THIS_SCRIPT_PATH/tsconfig.json"
```

**Review by mcdurdin — APPROVED** [review]
LGTM, just minor tweaks

**Comment by jahorton** [comment]
Accidentally skipped a rung when merging ancestor branches; that's what confused the labeler.

**Comment by darcywong00** [comment]
CI build failed at the "clean artifact directory" step

**Comment by jahorton** [comment]
> CI build failed at the "clean artifact directory" step

Yep, that's expected.  Things are a bit unstable at this point in the PR ladder.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.41-alpha](https://keyman.com/downloads/releases/alpha/17.0.41)
